src/libpmemfile/vfd_table.c

```diff
@@ -0,0 +1,663 @@
+	old_cwd_entry = cwd_entry;
+	cwd_entry = new_cwd_entry;
+
+	unref_entry(old_cwd_entry);
+}
+
+/*
+ * pmemfile_vfd_chdir_pf -- change the current working directory to
+ * a pmemfile handled directory.
+ */
+long
+pmemfile_vfd_chdir_pf(struct pool_description *pool, struct pmemfile_file *file)
+{
+	long result;
+
+	util_mutex_lock(&vfd_table_mutex);
+
+	if (pmemfile_fchdir(pool->pool, file) != 0) {
+		result = -errno;
+	} else {
+		struct vfile_description *entry = fetch_free_file_slot();
+
+		*entry = (struct vfile_description) {
+			.pool = pool, .file = file,
+			.kernel_cwd_fd = -1,
+			.is_special_cwd_desc = false,
+			.ref_count = 1};
+
+		chdir_exchange_entry(entry);
+
+		result = 0;
+	}
+
+	util_mutex_unlock(&vfd_table_mutex);
+
+	return result;
+}
+
+/*
+ * pmemfile_vfd_chdir_kernel_fd -- change the current working directory to
+ * a kernel handled directory.
+ */
+long
+pmemfile_vfd_chdir_kernel_fd(int fd)
+{
+	long result;
+
+	util_mutex_lock(&vfd_table_mutex);
+
+	result = syscall_no_intercept(SYS_fchdir, fd);
+
+	if (result == 0) {
+		struct vfile_description *entry = fetch_free_file_slot();
+
+		*entry = (struct vfile_description) {
+			.pool = NULL, .file = NULL,
+			.kernel_cwd_fd = fd,
+			.is_special_cwd_desc = true,
+			.ref_count = 1};
+
+		chdir_exchange_entry(entry);
+	}
+
+	util_mutex_unlock(&vfd_table_mutex);
+
+	return result;
+}
+
+static bool is_memfd_syscall_available;
+
+#ifdef SYS_memfd_create
+
+static void
+check_memfd_syscall(void)
+{
+	long fd = syscall_no_intercept(SYS_memfd_create, "check", 0);
+	if (fd >= 0) {
+		is_memfd_syscall_available = true;
+		syscall_no_intercept(SYS_close, fd);
+	}
+}
+
+#else
+
+#define SYS_memfd_create 0
+#define check_memfd_syscall()
+
+#endif
+
+/*
+ * acquire_new_fd - grab a new file descriptor from the kernel
+ */
+static long
+acquire_new_fd(const char *path)
+{
+	long fd;
+
+	if (is_memfd_syscall_available) {
+		fd = syscall_no_intercept(SYS_memfd_create, path, 0);
+		/* memfd_create can fail for too long name */
+		if (fd < 0) {
+			fd = syscall_no_intercept(SYS_open, "/dev/null",
+					O_RDONLY);
+		}
+	} else {
+		fd = syscall_no_intercept(SYS_open, "/dev/null", O_RDONLY);
+	}
+
+	if (fd >= (long)ARRAY_SIZE(vfd_table)) {
+		syscall_no_intercept(SYS_close, fd);
+		return -ENFILE;
+	}
+
+	return fd;
+}
+
+/*
+ * pmemfile_vfd_assign -- return an fd that can be used by an application
+ * in the future to refer to the given pmemfile file.
+ *
+ * Creates a new vfile_description entry with ref_count = one.
+ */
+long
+pmemfile_vfd_assign(struct pool_description *pool,
+			struct pmemfile_file *file,
+			const char *path)
+{
+	long new_vfd = acquire_new_fd(path);
+
+	if (new_vfd < 0)
+		return new_vfd;
+
+	struct vfile_description *entry = fetch_free_file_slot();
+
+	*entry = (struct vfile_description) {
+		.pool = pool, .file = file,
+		.kernel_cwd_fd = -1,
+		.is_special_cwd_desc = false,
+		.ref_count = 1};
+
+	util_mutex_lock(&vfd_table_mutex);
+
+	vfd_table[new_vfd] = entry;
```

Comment:
Well, we might check it, yes.
This is a new fd from the kernel, so that would mean either that pmemfile_vfd_close is buggy, or the kernel is buggy.